File: elements/src/main/java/com/rainmore/domains/Account.java
package com.rainmore.domains;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.time.LocalDateTime;

@Entity
@Table(name = "accounts")
public class Account implements Model {

    private Long id;
    private String firstName;
    private String lastName;
    private String email;
    private Boolean isActive;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;

    @PrePersist
    public void prePersist() {
        if (createdAt == null) createdAt = LocalDateTime.now();
        if (updatedAt == null) updatedAt = createdAt;

        this.preUpdate();
    }

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    @Id
    @Column
    @GeneratedValue(strategy = GenerationType.AUTO)
    public Long getId() {
        return id;
    }

    public Account setId(Long id) {
        this.id = id;
        return this;
    }

    @Column
    @NotNull
    public String getFirstName() {
        return firstName;
    }

    public Account setFirstName(String firstName) {
        this.firstName = firstName;
        return this;
    }

    @Column
    @NotNull
    public String getLastName() {
        return lastName;
    }

    public Account setLastName(String lastName) {
        this.lastName = lastName;
        return this;
    }

    @Column
    @NotNull
    public String getEmail() {
        return email;
    }

    public Account setEmail(String email) {
        this.email = email;
        return this;
    }

    @NotNull
    public Boolean getActive() {
        return isActive;
    }

    public Account setActive(Boolean active) {
        isActive = active;
        return this;
    }

    @Column(updatable = false)
    @NotNull
    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public Account setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
        return this;
    }

    @NotNull
    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public Account setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
        return this;
    }
}

File: elements/src/main/kotlin/com/rainmore/domains/Account.kt
package com.rainmore.domains

import java.time.LocalDateTime
import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
@Table(name = "accounts")
class Account : Model {
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "id", nullable = false)
    @Id
    var id: Long? = null

    @Column
    var firstName: String? = null

    @Column
    var lastName: String? = null

    @Column
    var email: String? = null

    @Column
    var active: Boolean? = null

    @Column(updatable = false)
    @NotNull
    var createdAt: LocalDateTime? = null

    @Column
    @NotNull
    var updatedAt: LocalDateTime? = null

    @PrePersist
    fun prePersist() {
        if (createdAt == null) createdAt = LocalDateTime.now()
        if (updatedAt == null) updatedAt = createdAt
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = LocalDateTime.now()
    }
}